File: Kawai codes/03.Integrals/ch03pr01.py
# -*- coding: utf-8 -*-
"""
%**************************************************************************
%*  Example  3.1                                                          *
%*  filename: ch03pr01.py                                                 *
%*  program listing number: 3.1                                           *
%*                                                                        *
%*  This program integrate sin(x) from x=0 to x=pi using rectangular,     *
%*  trapezoidal and simpson methods.  Absolute errors are plotted.        *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  01/14/2019.                                    *
%**************************************************************************
"""

import numpy as np
import scipy.integrate as integrate
import matplotlib.pyplot as plt

opt=input("Use SciPy [y/n] ")
if opt=='y':
    print("SciPy integrate will be used.\n")
else:
    print("SciPy will not be used.\n")
    
# Set the lower and upper bound of the integration
a=0.
b=np.pi/2.
# Header of the output
print("{0:^75}".format('Absolute error in various numrical integration'))
print("{0:^6} {1:^23} {2:^24} {3:^24} \n"
      .format('N','Rectangular','Trapezoidal','Simpson'))

kmax=10
h=np.zeros(kmax+1)
err_rect=np.zeros(kmax+1)
err_trap=np.zeros(kmax+1)
err_simp=np.zeros(kmax+1)

for k in range(0,kmax):
    N=2**(k+1)
    h[k]=(b-a)/N

    x = a + np.linspace(a,b,N+1)
    f = np.sin(x)
    
    rect=f[0:N].sum()*h[k]

    if opt=='y':
        trap=integrate.trapz(f,x)
        simp=integrate.simps(f,x)
    else:      
        trap=f[1:N].sum()*h[k]+(f[0]+f[N])*h[k]/2.    
        simp=(2.0*f[0:N-1:2].sum()+4.0*f[1:N:2].sum()-f[0]+f[N])*h[k]/3. 
    
    err_rect[k]=abs(1.-rect)
    err_trap[k]=abs(1.-trap)
    err_simp[k]=abs(1.-simp)

    print("{0:5d} {1:24.16e} {2:24.16e} {3:24.16e}"
          .format(N,err_rect[k],err_trap[k],err_simp[k]))
    
    del x
    del f
    
# Plot data
h2=h**2
h3=h**3
h4=h**4

plt.ioff()
plt.figure(figsize=(6,5))
plt.loglog(h,err_rect, 'og', label='rectangular')
plt.loglog(h,err_trap, 'ob', label='trapezoidal')
plt.loglog(h,err_simp, 'or', label='simpson')
plt.loglog(h,h,'--g',label='$h$')
plt.loglog(h,h2,'--b',label='$h^2$')
plt.loglog(h,h4,'--r',label='$h^4$')
plt.legend(loc=4)
plt.xlabel('h')
plt.ylabel('Integral')
plt.show()

File: Kawai codes/03.Integrals/ch03pr01.m
%**************************************************************************
%*     Example  3.1                                                       *
%*     filename: ch03pr01.m                                               *
%*     program listing number: 3.1                                        *
%*                                                                        *
%*  This program integrate sin(x) from x=0 to x=pi using rectangular,     *
%*  trapezoidal and simpson methods.  Absolute errors are plotted.        *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  01/06/2014.                                    *
%**************************************************************************

clear all;

% Set the lower and upper bound of the integration
a=0;
b=pi;

opt=input('Use MATLAB built-in integration? [y/n]','s');
if opt=='y'
    display('MATLAB built-in trapz will be used.');
else
    display('No MATLAB built-in function will be used.');
end

% Header of the output
display('    Absolute error in various numrical integration')
fprintf(' %3s %21s %24s %24s\n','N','Rectangular','Trapezoidal','Simpson');

% loop over different N
for k=1:10
    N=2^k;
    h(k)=(b-a)/N;
    
    % evaluate the variable and function
    for i=0:N
         x(i+1)=a+i*h(k);
         f(i+1)=sin(x(i+1));
    end
    
    % Rectangular rule
    rect=sum(f(1:N))*h(k);
    
    % Trapezoidal rule
    if opt=='y'
        % Use MATLAB trapz
        trap=trapz(x,f);
    else
        % Manual
        trap=sum(f(2:N))*h(k) + (f(1)+f(N+1))*h(k)/2;
    end
    
    % Simpson rule
    simp=(2*sum(f(1:2:N-1))+4*sum(f(2:2:N))-f(1)+f(N+1))*h(k)/3;

    % Evaluation of errors (the exat answer is 1)
    err_rect(k)=abs(1-rect);
    err_trap(k)=abs(1-trap);
    err_simp(k)=abs(1-simp);
    
    %print out the results
    fprintf('   %5d %24.16e %24.16e %24.16e \n',...
        N,err_rect(k),err_trap(k),err_simp(k));
end

% Order of the errors
h2=h.^2;
h3=h.^3;
h4=h.^4;

% Plot the results
subplot(1,1,1)
p=loglog(h,err_rect,'d',h,err_trap,'o',h,err_simp,'s',...
    h,h,'--',h,h2,'--',h,h4,'--');

% Format the plot
title('Absolute errors in the various integration methods');
xlabel('h');
ylabel('absolute error');
set(p(1),'Color','green');
set(p(2),'Color','blue');
set(p(3),'Color','red');
set(p(4),'Color','green','LineWidth',2);
set(p(5),'Color','blue','LineWidth',2);
set(p(6),'Color','red','LineWidth',2);
legend(p,{'rectangular','trapezoidal','simpson','h','h^2','h^4'});
legend(p,'Location','SouthEast');
